Replace hash maps and sets with `indexmap` equivalents

sug0 opened the pull request:
## Describe your changes

Closes #2576

## Indicate on which release or other PRs this topic is based on

Based on #2682

## Checklist before merging to `draft`
- [ ] I have added a changelog
- [ ] Git history is in acceptable state

sug0 commented:
might need to rebuild wasms for tests

sug0 commented:
**TODO:** update indexmap
https://github.com/heliaxdev/indexmap/releases/tag/2.2.3-heliax-3

cwgoes reviewed (dismissed):
ACK modulo CI

sug0 commented:
e2e tests in CI passed without rebuilding wasms for tests, but it might still be wise to rebuild them anyway.

tzemanovic commented:
I've added [f1e4d72](https://github.com/anoma/namada/pull/2685/commits/f1e4d72151cad9e9e5b56a410c35bed8555733a6) to prevent from using them again

sug0 commented:
updated the `indexmap` dep

sug0 commented:
rebased on `v0.31.8` and updated `indexmap` to `2.2.4-heliax-1` (the new upstream version prevents OOM errors in the `serde::Deserialize` implementation)
